File: algorithm/86.partition-list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        
        if head is None:
            return None

        left_head = ListNode(0)
        left_head.next = head

        right_head = ListNode(0)

        left = left_head
        right = right_head

        while head:

            if head.val < x:
                left.next = head
                left = left.next
            else:
                right.next = head
                right = right.next

            head = head.next

        right.next = None
        left.next = right_head.next

        return left_head.next

File: leetcode/86.partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        
        ListNode newList(0);
        ListNode *new_cur = &newList;
        
        ListNode oldList(0);
        oldList.next = head;
        ListNode *cur = &oldList;
        
        while(cur && cur->next)
        {
            if (cur->next->val < x)
            {
                ListNode *tmp = cur->next;
                cur->next = tmp->next;
                
                // append to new list
                tmp->next = NULL;
                new_cur->next = tmp;
                new_cur = new_cur->next;
            }
            else
            {
                cur = cur->next;
            }          
        }
        
        new_cur->next = oldList.next;
        
        return newList.next;
    }
};